mesonbuild/backend/vs2010backend.py

```diff
@@ -95,18 +98,59 @@ def split_o_flags_args(args: T.List[str]) -> T.List[str]:
             o_flags += ['/O' + f for f in flags]
     return o_flags
 
-
 def generate_guid_from_path(path, path_type) -> str:
     return str(uuid.uuid5(uuid.NAMESPACE_URL, 'meson-vs-' + path_type + ':' + str(path))).upper()
 
 def detect_microsoft_gdk(platform: str) -> bool:
     return re.match(r'Gaming\.(Desktop|Xbox.XboxOne|Xbox.Scarlett)\.x64', platform, re.IGNORECASE)
 
+def filtered_src_langs_generator(sources: T.List[str]):
+    for src in sources:
+        ext = src.split('.')[-1]
+        if compilers.compilers.is_source_suffix(ext):
+            yield compilers.compilers.SUFFIX_TO_LANG[ext]
+
+# Returns the source language (i.e. a key from 'lang_suffixes') of the most frequent source language in the given
+# list of sources.
+# We choose the most frequent language as 'primary' because it means the most sources in a target/project can
+# simply refer to the project's shared intellisense define and include fields, rather than have to fill out their
+# own duplicate full set of defines/includes/opts intellisense fields.  All of which helps keep the vcxproj file
+# size down.
+def get_primary_source_lang(target_sources: T.List[File], custom_sources: T.List[str]) -> T.Optional[str]:
+    lang_counts = Counter([compilers.compilers.SUFFIX_TO_LANG[src.suffix] for src in target_sources if compilers.compilers.is_source_suffix(src.suffix)])
+    lang_counts += Counter(filtered_src_langs_generator(custom_sources))
+    most_common_lang_list = lang_counts.most_common(1)
+    # It may be possible that we have a target with no actual src files of interest (e.g. a generator target),
+    # leaving us with an empty list, which we should handle -
+    return most_common_lang_list[0][0] if most_common_lang_list else None
+
+# Returns a dictionary (by [src type][build type]) that contains a tuple of -
+# (pre-processor defines, include paths, additional compiler options)
+# fields to use to fill in the respective intellisense fields of sources that can't simply
+# reference and re-use the shared 'primary' language intellisense fields of the vcxproj.
+def get_non_primary_lang_intellisense_fields(
```

Comment:
Any particular reason these are freestanding functions as opposed to methods on the VS backend?

Reply:
I can suggest a few reasons why it might be preferable as a free function (over a method) -
- Making it a class method _can_ make a reader incorrectly assume there's the possibility for affecting or interacting with intermal member state.
- Conversely, a free function, particularly one that has no need for any implicit member data/functionality (which is the case with this function) just makes it crystal clear to the reader what the inputs are and that there's no hidden, implicit inputs to muddy the reader's understanding of what's going on in the func.
- Following on from the point above, free funcs are very often easier (and at least never harder) to refactor (split, merge, relocate) than member funcs, and to reason about while refactoring.
- The function, in this case, doesn't need to use any members of self/Vs2010Backend.

Reply:
Having them as freestanding functions is fine. There is a Python code checker (I forget which) that actually complains if you have a method that does not use its `self` argument at all.